### DISCUSSION TITLE
My team have a code review for using your json library, but the rules are really......

### DISCUSSION THREAD

**Discussion by ly5519** [nlohmann/json#4357] [Q&A]
our object no exception, so must be checked before use, like this:
```C++
nlohmann::json json;
if (json[KEY_DEV_ID].is_string()) {
    std::string str = json[KEY_DEV_ID].get<std::string>();
    dosomething();
    // ......
}
```

but rules say I should:
```C++
nlohmann::json json;
if (json.contains(KEY_DEV_ID) && json[KEY_DEV_ID].is_string()) {
    std::string str = json[KEY_DEV_ID].get<std::string>();
    dosomething();
    // ......
}
```
this may look not bad, how about this?🤣
```C++
nlohmann::json json;
if (json.contains(KEY_DATA) && json[KEY_DATA].is_object() &&
    json[KEY_DATA].contains(HANDSHANKE_DATA) && json[KEY_DATA][HANDSHANKE_DATA].is_object() &&
    json[KEY_DATA][HANDSHANKE_DATA].contains(KEY_DEV_ID) && json[KEY_DATA][HANDSHANKE_DATA][KEY_DEV_ID].is_string()) {
    std::string str = json[KEY_DATA][HANDSHANKE_DATA][KEY_DEV_ID].get<std::string>();
    dosomething();
    // ......
}
```
I think only need the last one, like this:
```C++
nlohmann::json json;
if (json[KEY_DATA][HANDSHANKE_DATA][KEY_DEV_ID].is_string()) {
    std::string str = json[KEY_DATA][HANDSHANKE_DATA][KEY_DEV_ID].get<std::string>();
    dosomething();
    // ......
}
```

Is it right, is it really necessary?

help me！thank you😘

**Comment by dingminghbut** [comment]
Your reviewer is right about the important part: the last version is not safe.

The reason is that non-const `operator[]` is not only a read operation. If a key does not exist, it can insert a `null` value into the object while you are just checking it. So this:

```cpp
if (json[KEY_DATA][HANDSHANKE_DATA][KEY_DEV_ID].is_string()) {
    // ...
}
```

can mutate `json` by creating missing intermediate values. It also makes it harder to distinguish "missing key" from "present but null/wrong type".

For a simple top-level key, this is fine:

```cpp
if (json.contains(KEY_DEV_ID) && json[KEY_DEV_ID].is_string()) {
    auto str = json[KEY_DEV_ID].get<std::string>();
}
```

For deeply nested paths, I would avoid repeating every level manually and use a JSON pointer plus `contains()` / `at()`:

```cpp
const nlohmann::json::json_pointer ptr("/data/handshake/dev_id");

if (json.contains(ptr)) {
    const auto& value = json.at(ptr); // no insertion
    if (value.is_string()) {
        auto str = value.get<std::string>();
        dosomething();
    }
}
```

That keeps the two important properties:

- no accidental insertion while checking;
- no exception from `at()` because you only call it after `contains(ptr)`.

If your keys can contain `/` or `~`, remember that JSON Pointer requires escaping them as `~1` and `~0`.

So yes, the checks are necessary if you want no exceptions and no accidental mutation. But you can make them much cleaner by wrapping the lookup in a small helper or using `json_pointer` instead of spelling out every level.

**Reply by ly5519** [reply]
thank you very much👍
